Refactor classic menu conversion process

getdave opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/repository-management.md#pull-requests. -->

<!-- Gutenberg's license is in the process of updating to be dual-licensed under the GPL and MPL. As part of that transition, all new contributions are dual-licensed. For more information, see: https://github.com/WordPress/gutenberg/blob/trunk/LICENSE.md -->

## Description
Alternative approach to https://github.com/WordPress/gutenberg/pull/38824. Continues working towards addressing https://github.com/WordPress/gutenberg/issues/37190.

This PR seeks both to

- simplify the logic around converting classic menus to block based menus in the Nav block.
- avoid the impression of the UI "locking up" when the classic menu is selected and instead provide feedback to the user that an operation is taking place.

The current hooks are complex to reason about and don't expose enough information about the progress of the conversion to allow for the UI to respond (e.g. loading states...etc).

This PR builds on the approach outlined by @talldan in https://github.com/WordPress/gutenberg/pull/38824#pullrequestreview-883918504. Essentially we create a new hook which exposes:

1. A function which trigger the conversion process.
2. The state of the conversion process.

~🙏 **Please note**: This PR _does not_ attempt to solve the loading state indicator problems that are noticeable in this PR. This is being addressed in https://github.com/WordPress/gutenberg/pull/38907.~

## Outstanding problems to solve

- [x] Allow the dropdown to close immediately upon click to avoid the UI feeling like it's locking up.
- [x] ~Create loading UI states _in the block_ whilst the conversion process happens (good things we expose the state!).~ To be addressed following merge of https://github.com/WordPress/gutenberg/pull/38907.
- [x] Avoid component unmount errors when closing the dropdown whilst the async conversion process is still running.
- [x] Migrate towards usage of registry as suggest by @adamziel in https://github.com/WordPress/gutenberg/pull/38824#issuecomment-1041549542.
- [x] Handle errors from the async conversion process (`.catch()` .etc).
- [x] Provide accessible progress announcements of conversion status.

I will tackle those steps during the remainder of this week.


## Testing Instructions

1. Switch to a classic Theme and create a Menu using the `Menus` screen. Add lots of items just for fun.
2. Create another Menu with no items in it. Be sure to save.
3. Switch to block Theme.
4. New Post.
5. Insert a Navigation block.
6. Select the populated classic menu.
7. Immediately see a loading state appear and not sense that the UI is "locking up" (for an understand of what I mean check out the same proceedure on `trunk` or watch the screencaptures below).
8. See the menu converted successfully and inserted into the block.
9. Repeat the same process but this time using the unpopulated classic menu. You should still see the conversion happen but the block will end up being empty because the classic menu was empty. That's expected.
10. Try and test what happens if you go offline just before you click on the classic menu. The UI should handle things and you should see an error notice pop up.
11. Try using screen reader software to run the above steps. You should be able to perceive the progress of the conversion without having to look at the screen. It should be announced to you.

## Screenshots <!-- if applicable -->

### Before

Notice how the dropdown doesn't close and  UI seems to lock up/freeze when I click the classic menu. 

https://user-images.githubusercontent.com/444434/154458859-495a22f5-4138-4dd9-87b1-55ca50706f2e.mp4



### After

Notice the UI lock up is gone. The dropdown disappears and we see the block loading state immediately which resolves as appropriate:

https://user-images.githubusercontent.com/444434/156602370-9478c9fe-c352-4bab-b137-affc5237ebad.mp4

#### Error handling

Notice that the UI communicates the error if the request fails:

https://user-images.githubusercontent.com/444434/156608077-385a689d-ebe2-43fc-a2f3-aa9ed90571d2.mov



## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->
- [ ] I've updated related schemas if appropriate. <!-- Reference: https://github.com/WordPress/gutenberg/tree/trunk/schemas -->

adamziel commented on packages/block-library/src/navigation/edit/use-convert-classic-menu-to-block-menu.js:
```diff
@@ -0,0 +1,124 @@
+
+	const safeDispatch = useSafeDispatch( dispatch );
+
+	async function convertClassicMenuToBlockMenu( menuId, menuName ) {
```
I'd pull it out of the hook, there's no need to construct a new function on each run and it would be more readable to me (more separation).

getdave replied:
I take your point but if you follow the flow you'll notice that this is required to be within a hook because it uses other hooks. In fact it's hooks within hooks within hooks.

I'm not in favour of this approach really but I don't think we should look to refactor in this PR because it will become overly complex to review.

How would you feel about pairing on a refactor of this sometime?

adamziel replied:
Sure, let's try to pair up tomorrow

adamziel replied:
With the recent updates in place I feel pretty good about having this here, let's just wrap it in `useCallback`.

adamziel commented on packages/block-library/src/navigation/edit/use-convert-classic-menu-to-block-menu.js:
```diff
@@ -0,0 +1,124 @@
+	}
+}
+
+function useSafeDispatch( dispatch ) {
```
Hey that's smart! I'd love to see a PR to `@wordpress/data` with this change.

adamziel replied:
Just to make sure I understand what's going on – do you want to avoid errors when the action is dispatched by an unmounted component?

getdave replied:
Yes. The "safe" aspect stops the dispatch if the component is unmounted. I will propose this in a core package in a separate PR/

getdave replied:
Just noting I proposed this but it was rejected https://github.com/WordPress/gutenberg/pull/38956

adamziel commented on packages/block-library/src/navigation/edit/use-convert-classic-menu-to-block-menu.js:
```diff
@@ -0,0 +1,124 @@
+
+	const convert = useCallback(
+		( menuId, menuName ) => {
+			if ( ! menuId || ! menuName ) {
```
You can have an async function inside useCallback, only you have to declare it separately, e.g.:

```js
const convert = useCallback(
	( menuId, menuName ) => {
		async function doTheWork() {};
		doTheWork();
	}
);
```

getdave replied:
Yeh but in this case the aim isn't to return `Promise` from `convert()`. We want to run a routine and then have that routine update the `state` which we can then use outside of the hook to decide on various things such as loading states.

This relieves the hook consumer from having to reimplementing the `isFetching`, `isError`...etc inside their component. Rather all this is encapsulated and we simply expose a state which the consumer can use to decide on side effects within a `useEffect`.

adamziel replied:
All I'm saying is there are two code styles available to choose from. Note that my snippet above doesn't return a promise either.

getdave replied:
Oh I see. You're just saying we could move the async function outside of the `useCallback` and that would be ok.

Unless it's a blocker for you, I'd rather not do that now because I'm planning a follow up that might change things a little. I ping you when that's up.

adamziel commented on packages/block-library/src/navigation/edit/use-convert-classic-menu-to-block-menu.js:
```diff
@@ -0,0 +1,124 @@
+	const createNavigationMenu = useCreateNavigationMenu( clientId );
+	const registry = useRegistry();
+
+	const [ state, dispatch ] = useReducer( reducer, {
+		navMenu: null,
+		isFetching: false,
+	} );
```
That seems a bit overcomplicated to me for two state variables and three possible transitions, what do you think about the following?

```js
const [ navMenu, setNavMenu ] = useState( null );
const [ isFetching, setIsFetching ] = useState( false );
const setLoading = useCallback( () => {
	setIsFetching( true );
	setNavMenu( null );
} );
const setError = useCallback( () => {
	setIsFetching( false );
	setNavMenu( null );
} );
const setResolved = useCallback( ( menus ) => {
	setIsFetching( false );
	setNavMenu( menus );
} );
```

getdave replied:
Just noting I proposed this but it was rejected https://github.com/WordPress/gutenberg/pull/38956

getdave replied:
I just felt the reducer was cleaner and easier to reason about. Moreover, if we want to avoid the set state on unmount error we're going to have to call the `useSafeDispatch` once for each `setState` in order to get the wrapped version of that handler.

adamziel replied:
That error seems to be harmless and will disappear in the next React release though?

getdave replied:
I've had a change of heart. Let's avoid adding complexity where it's not warranted. I've refactored to `useState`.

getdave commented:
I will resume work on this PR once https://github.com/WordPress/gutenberg/pull/38907 is merged.

getdave committed:
Refactor hook to single async function

getdave committed:
Adjust UI state vars to account for conversion process state

getdave commented on packages/block-library/src/navigation/edit/index.js:
```diff
@@ -310,6 +325,15 @@ function Navigation( {
+			! classicMenuConversionState?.isFetching &&
+			classicMenuConversionState.navMenu
+		) {
+			setRef( classicMenuConversionState.navMenu?.id );
```
This is confusing. Due to the `useState` var name containing the term `ref` It could appear we are setting a `React.ref` here which would not cause the component to re-render and thus the UI would not reflect the new menu ID.

In actual fact this is a `setState` call.

We should look to refactor this `ref` term away to something else.

getdave committed:
Unify menu ref update process

getdave committed:
Remove superceeded hook

getdave committed:
Refactor conversion state to enum

getdave committed:
Add initial error handling for conversion process

getdave committed:
i18n of classic menu fetch error

getdave committed:
Handle more creation errors

getdave committed:
Manually hoist var

getdave commented on packages/block-library/src/navigation/edit/index.js:
```diff
@@ -231,23 +232,38 @@ function Navigation( {
+		state: classicMenuConversionState,
+	} = useConvertClassicToBlockMenu( clientId );
+
+	const isConvertingClassicMenu =
```
Maybe the exposed state should be "converted" rather than "fetching". It would be more descriptive and we could do away with this var.

getdave commented on packages/block-library/src/navigation/edit/index.js:
```diff
@@ -231,23 +232,38 @@ function Navigation( {
 	const isPlaceholder =
-		! ref && ( ! hasUncontrolledInnerBlocks || isWithinUnassignedArea );
+		! ref &&
+		! isConvertingClassicMenu &&
```
No showing a placeholder if we're running the conversion process.

getdave commented on packages/block-library/src/navigation/edit/index.js:
```diff
@@ -231,23 +232,38 @@ function Navigation( {
 	// - the Navigation Post isn't available (hasn't resolved) yet.
-	const isLoading = !! ( ref && ! isEntityAvailable );
+	const isLoading =
+		isConvertingClassicMenu ||
```
Now `loading` UI state should show _immediately_ if we are running the conversion.

getdave commented on packages/block-library/src/navigation/edit/index.js:
```diff
@@ -490,11 +534,18 @@ function Navigation( {
+							return;
+						}
+
+						if ( isClassicMenu ) {
+							// Setting ref and selecting block
+							// handled by effect post-conversion.
+							convert( post.id, post.name );
+						} else {
+							handleUpdateMenu( post.id );
```
I don't love this. Open to ideas about how to improve.

getdave commented on packages/block-library/src/navigation/edit/index.js:
```diff
@@ -513,6 +564,9 @@ function Navigation( {
 								onSelect={ ( { id } ) => {
 									setRef( id );
 								} }
+								onSelectClassic={ ( classicMenu ) => {
```
Alternative to a separate `onSelectClassic` would be to detect whether it's a classic menu that's been selected by looking for a property of a classic menu that isn't present on a Navigation Menu (if there is one).

getdave replied:
I've gone with the approach of detecting the classic menu by checking for the rather obscure `auto_add` prop.

getdave commented on packages/block-library/src/navigation/edit/index.js:
```diff
@@ -513,6 +564,9 @@ function Navigation( {
 									setRef( id );
 								} }
+								onSelectClassic={ ( classicMenu ) => {
+									convert( classicMenu.id, classicMenu.name );
```
This kicks of the async process but is non blocking and allows the UI of the block to update independently based on the conversion status provided by the hook.

adamziel replied:
If it made sense to say that in the PR, it makes sense to say it in a comment in that file

getdave commented on packages/block-library/src/navigation/edit/use-convert-classic-menu-to-block-menu.js:
```diff
@@ -0,0 +1,168 @@
+						navMenu,
+					} );
+				} )
+				.catch( ( e ) => {
+					safeDispatch( {
+						type: 'ERROR',
+						error: e?.message,
+					} );
+				} );
```
Centralised error handling here.

All errors are bubbled up to the UI via state changes on the reducer, which are then exposed via the hook's `state` property.

The consuming component can read `state.error` to get the error and `state.status` to determine if there was an error.

Note that `convertClassicMenuToBlockMenu` implements specialised error handling to ensure that error messages are descriptive of what has occurred (i.e. we don't rely on default messages coming from `core-data`...etc.

getdave commented on packages/block-library/src/navigation/edit/use-navigation-notice.js:
```diff
@@ -5,19 +5,19 @@ import { useRef } from '@wordpress/element';
 	const { createWarningNotice, removeNotice } = useDispatch( noticeStore );
 
-	const showNotice = () => {
+	const showNotice = ( customMsg ) => {
```
Allow for an optional custom message at call time

getdave commented:
@jasmussen I wonder if you would take this for a spin on the design side?

getdave committed:
Add accessible menu import announcements

getdave commented:
Just spun this one out to capture the need to [avoid duplicate classic menu names during import](https://github.com/WordPress/gutenberg/issues/39200).

getdave commented:
Accessibility review required just to validate the usage of `speak()` here is appropriate. I tested with voiceover and it seemed to provide a good level of feedback during the classic menu conversion process whereas before there was absolutely none.

getdave committed:
Refactor to remove useReducer in favour of simple useState

Addresses https://github.com/WordPress/gutenberg/pull/38858#discussion_r808956065

getdave committed:
Fix bug where constants were accidentally assigned to constants

getdave committed:
Simplify menu selector props

Detect the classic menu by relying on it’s `auto_add` prop which we know will not be defined on a Navigation Post.

scruffian commented on packages/block-library/src/navigation/edit/index.js:
```diff
@@ -310,6 +328,42 @@ function Navigation( {
+
+	useEffect( () => {
+		if ( classicMenuConversionStatus === 'pending' ) {
+			speak( __( 'Classic menu importing.' ) );
```
These messages are pretty helpful, should we also add them as snackbar notifications? (In a different PR)

getdave replied:
We could yes if that's not too "noisy"? I wonder if notification popups have `speak()` built in?

getdave replied:
Actually I've found a great case where that would be very useful. When creating an empty menu it's not clear that it's been "successful". An explicit confirmation would be really helpful.

I'm working on a sister PR now which I will push up which refactors the create empty process.

scruffian reviewed (approved):
This works as described. The code looks good to me, but I'm not familiar with it.

adamziel commented on packages/block-library/src/navigation/edit/use-convert-classic-menu-to-block-menu.js:
```diff
@@ -0,0 +1,122 @@
+				__( `Unable to fetch classic menu "%s" from API.` ),
+				menuName
+			),
+			{
+				menuId,
+				menuName,
+			}
```
I think this information is lost. [The second argument to Error is `options`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error/Error#rethrowing_an_error_with_a_cause) and the only acceptable option is `cause`. You could wrap that in `cause` as follows:

```suggestion
			{
				cause: {
					menuId,
					menuName,
				}
			}
```

But it doesn't seem like we're actually doing anything with this information so we could just remove it entirely:

```suggestion
```

Also, I believe the stack trace points to the line where the error is constructed, not thrown. How about reusing a variable called `errorMessage = sprintf( // ...` instead of `fetchError`?

getdave replied:
I've ended up inlining everything and passing the original error as `cause`.

adamziel commented on packages/block-library/src/navigation/edit/use-convert-classic-menu-to-block-menu.js:
```diff
@@ -0,0 +1,122 @@
+					__( `Unable to create Navigation Menu "%s".` ),
+					menuName
+				),
+				{
+					menuId,
+					menuName,
+				}
```
This could be deleted, too

adamziel replied:
Or better, replaced with { cause: e }

getdave replied:
I did see this `cause` and was thinking "I'll come back to that". Guess what...I forgot! Lesson - don't leave issues/problems undocumented as you come across them - you will forget 😄 

Thanks for catching the silly usage of `options`. I've now updated to pass the original error as `cause`.

adamziel commented on packages/block-library/src/navigation/edit/use-convert-classic-menu-to-block-menu.js:
```diff
@@ -0,0 +1,122 @@
+		let navigationMenu;
+		let classicMenuItems;
+
+		const menuItemsParameters = {
```
This is only used once – let's inline it.

getdave replied:
`menuItemsParameters` is now inlined.

adamziel commented on packages/block-library/src/navigation/edit/use-convert-classic-menu-to-block-menu.js:
```diff
@@ -0,0 +1,122 @@
+				.resolveSelect( coreStore )
+				.getMenuItems( menuItemsParameters );
+		} catch ( e ) {
+			throw fetchError;
```
We lose the error information here. How about `throw new Error( fetchMessage, { cause: e } );`?

getdave replied:
✅ I've rethrown the error and avoided overthinking error abstractions.

adamziel commented on packages/block-library/src/navigation/edit/use-convert-classic-menu-to-block-menu.js:
```diff
@@ -0,0 +1,122 @@
+					setStatus( CLASSIC_MENU_CONVERSION_SUCCESS );
+				} )
+				.catch( ( e ) => {
+					setError( e?.message );
```
Let's re-throw this exception so that it can be debugged in devtools.

getdave replied:
Just checking...if we rethrow the error and it isn't handled then won't the block crash? I was trying to avoid that.

getdave replied:
Ok I tested and it's all fine. The block handles the error but the message is still displayed in dev tools. Win 👍

adamziel commented on packages/block-library/src/navigation/edit/index.js:
```diff
@@ -310,6 +328,42 @@ function Navigation( {
+	}
+
+	useEffect( () => {
+		if ( classicMenuConversionStatus === 'pending' ) {
```
We use `"pending"` here and `CLASSIC_MENU_CONVERSION_PENDING` in the other file – how about using these constants here as well? I don't like that long name, though. Ideally we could rely on TS to verify the right-hand side here.

getdave replied:
Yeh I was thinking that as well 😄 

AFAIK we aren't supposed to use TS at this level (Gutenberg informal agreement) but yes it would be ideal.

getdave replied:
✅

getdave commented:
Thanks for catching (excuse the pun) those lost errors @adamziel. Let me rework those.

getdave committed:
Improve error handling

getdave committed:
Use constants over comparisons for statuses

adamziel commented on packages/block-library/src/navigation/edit/use-convert-classic-menu-to-block-menu.js:
```diff
@@ -0,0 +1,135 @@
+					setError( err?.message );
+					setStatus( CLASSIC_MENU_CONVERSION_ERROR );
+
+					// Rethrow error for debugging.
```
nit, this could say just `throw err` since we're not going to tell the user what the error message says.

adamziel commented:
The only remaining thing I don't like is how many times it says _classic menu conversion_. I don't want to block on that, though, especially since I don't have any shorter naming scheme to propose.


**Edit** actually, I left a few more notes

adamziel commented on packages/block-library/src/navigation/edit/index.js:
```diff
@@ -490,11 +549,18 @@ function Navigation( {
-						if ( post ) {
-							setRef( post.id );
+					onFinish={ ( navPostOrClassicMenu ) => {
+						const isClassicMenu = navPostOrClassicMenu.hasOwnProperty(
```
Could that be a separate function? My reasoning:

* this logic is repeated in two places
* I'd rather have the future us find that function when we need it, than miss this very specific part of edit/index.js and invent another way of checking whether something is a classic menu

getdave replied:
Yes I agree. Looks like overtime the signatures of this handler become identical. That's good news as the incremental refactoring is obviously making the codebase easier to reason about.

I've extracted a common utility.

adamziel commented on packages/block-library/src/navigation/edit/index.js:
```diff
@@ -490,11 +549,18 @@ function Navigation( {
+								navPostOrClassicMenu.name
+							);
+						} else {
+							handleUpdateMenu( navPostOrClassicMenu?.id );
```
Can `navPostOrClassicMenu` be null? If yes, then the hasOwnProperty call above should take that into account.

getdave replied:
For now I've exited early. Perhaps we ought to consider showing a error, but I can't yet see a way where this would be `null` although it might happen.

adamziel commented on packages/block-library/src/navigation/edit/index.js:
```diff
@@ -510,8 +576,18 @@ function Navigation( {
-								onSelect={ ( { id } ) => {
-									setRef( id );
+								onSelect={ ( navPostOrClassicMenu ) => {
+									const isClassicMenu = navPostOrClassicMenu.hasOwnProperty(
```
Ditto the other two comments

getdave committed:
Extract common function for handling selecting a Navigation

getdave committed:
Optimise handler

getdave commented:
> The only remaining thing I don't like is how many times it says _classic menu conversion_

I agree. I think perhaps a var naming follow up is in order. Especially when you see which way https://github.com/WordPress/gutenberg/pull/39219 is heading.

adamziel commented on packages/block-library/src/navigation/edit/index.js:
```diff
@@ -388,6 +447,25 @@ function Navigation( {
+			if ( isClassicMenu ) {
+				convert( navPostOrClassicMenu.id, navPostOrClassicMenu.name );
+			} else {
+				handleUpdateMenu( navPostOrClassicMenu?.id );
```
nit: `?` not needed since we return early

getdave replied:
✅

adamziel commented on packages/block-library/src/navigation/edit/index.js:
```diff
@@ -388,6 +447,25 @@ function Navigation( {
+				return;
+			}
+
+			const isClassicMenu = navPostOrClassicMenu?.hasOwnProperty(
```
ditto

getdave replied:
✅

getdave committed:
Remove unwanted conditional operator

getdave commented:
I'm going to merge this once tests pass unless I get any more ❌ feedback.

jasmussen commented:
I saw the ping last week and did not have time to look at it. From a quick glance it looks good:

![nav](https://user-images.githubusercontent.com/1204802/157004038-d5fb86fc-1d00-4ec0-85e6-7cae0d01331c.gif)

The following issue is present in trunk, and subject to separate conversations, so not something that should block this PR. But showing the old menu even after it's being converted, as well as tools to manage them, it adds a fair bit of confusion: when do those tools appear? Why can I convert my menu again, wasn't it converted already? Worth us all looking at separately, but not precluded by these enhancements. Nice!

getdave commented:
> I saw the ping last week and did not have time to look at it. 

No worries.

> From a quick glance it looks good

👍 


> The following issue is present in trunk, and subject to separate conversations, so not something that should block this PR. But showing the old menu even after it's being converted, as well as tools to manage them, it adds a fair bit of confusion: when do those tools appear? Why can I convert my menu again, wasn't it converted already? Worth us all looking at separately, but not precluded by these enhancements. Nice!

Agreed. I feel like there's an Issue tracking this but I can't find it right now.
